fix: handle files with unspecified path in `getRulesMetaForResults`

fasttime opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[X] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

Fixes #16410

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

* In `getRulesMetaForResults`, files with an unspecified path are now treated as if they were located inside `cwd`.
* In `getRulesMetaForResults`, when a result referencing a rule has no config, we will explicitly throw an error with a descriptive message.
* Added top-level, internal functions `getPlaceholderPath` and `createExtraneousResultsError` to avoid code repetition.
* Added two new unit tests.
* Renamed an existing unit test to better disambiguate it from a new one. Also changed the assertion to check both error message and constructor.

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

mdjermanovic reviewed (approved):
LGTM, but would like @nzakas to take a look.

mdjermanovic commented:
Oh, and there's a merge conflict.

fasttime commented:
Thanks for reviewing @mdjermanovic! I've rebased the branch and fixed the merge conflict.

nzakas commented on lib/eslint/flat-eslint.js:
```diff
@@ -689,16 +710,20 @@ class FlatESLint {
+                 */
+                const config = configs.getConfig(filePath);
+
+                if (!config) {
```
There may be an edge case here. getConfig returns undefined if a file is ignore by the configuration, so it doesn’t necessarily mean that the file wasn’t run in this instance. I think this will result in an error if the results contain a warning for a file that was ignored.

You can test this by passing in a file that is specifically ignored.

fasttime replied:
Yes, I also thought about that. Now that #16402 has been fixed, the specific edge case you mention should be handled correctly. I will rebase the branch once again and add a new unit test to check the behavior in case an explicitly ignored file appears in the results when `warnIgnored` is set. Thanks for the feedback!

fasttime replied:
I've added a new unit test for when an anonymous file is ignored and linted with `warnIgnored`. A similar case, where the file has a specified path, is already covered in another test: https://github.com/eslint/eslint/blob/740b20826fadc5322ea5547c1ba41793944e571d/tests/lib/eslint/flat-eslint.js#L3864-L3867

nzakas replied:
I think we should add a test with `lintFiles()` as well. We need to make sure that if there are multiple files linted and one ignored that this won't throw an error.

My concern: `isFileIgnored()` just does `getConfig() === undefined`, so what you are really testing here is if a file is ignored. Because of that, it still seems like there's an edge case here that will bite us.

I think it would be safe to remove this check.

fasttime replied:
Ignored file results, as they don't have any messages with a `ruleId`, should be skipped by the `continue` statement a few lines above.

The point of checking if a config is found here is to avoid results linted by a different engine. If a result was not linted by the current engine, it may have no config without being ignored. This specific case is covered in [this test](https://github.com/eslint/eslint/pull/16437/files#diff-bef458c9c787f61228076dadf8878220be8c0f0ab489daa129a0d237a2feae17R3753-R3783). If this check is removed, `getRulesMetaForResults` will still fail, but with a non-specific error message (because `config.plugins` cannot be accessed at some point).

Adding a test with linted files and an ignored file at the same time is a good idea. There are no tests with multiple results so far. Thanks for the suggestion!

fasttime commented:
Rebased and updated as suggested by @nzakas. Since the new unit tests are only relevant to `FlatESLint`, I put them all at the start of the `describe` block, just under another existing, `FlatESLint`-specific unit test.

fasttime commented:
Added another unit test with `lintFiles()` as suggested by @nzakas.

fasttime commented on tests/lib/eslint/flat-eslint.js:
```diff
@@ -3813,6 +3813,26 @@ describe("FlatESLint", () => {
+            });
+
+            const results = await engine.lintFiles(["missing-semicolon.js", "passing.js", "undef.js"]);
+            const rulesMeta = engine.getRulesMetaForResults(results);
```
Okay, I've added an assertion to make sure that one of the messages returned by `lintFiles` belongs to an ignored file.
I chose this implementation because it doesn't require calling another method on the engine under test.
Alternatively, we could pass the `filePath` of each result to `isPathIgnored()`.

nzakas replied:
This works. 👍

nzakas reviewed (approved):
LGTM. Thanks so much for this.
